### PR TITLE
Test staking - Part 1

### PR DESCRIPTION
**Pull request by aakoshh**
Creates a `fendermint/testing/contract-test` crate with a `tests/stm_staking.rs` module for testing the staking actors against an idealised model.

- [x] separate test module for staking
- [x] `arbitrary::Arbitrary` instances to create test genesis 
- [x] initialise the state with some accounts and a subnet with some initial validators
- [x] definition of random operations
- [x] implement an idealised staking model

Deferred for a follow up PR:
- [ ] call the real contract for state transitions
- [ ] collect emitted events from contract calls
- [ ] validate emitted events
- [ ] validate balances 
- [ ] validate active powers

Changes unrelated to the test that made debugging easier:
* Added a `RegistryCaller` and `SubnetCaller` component in the testing crate to interact with the contracts
* All `ContractCaller` instances have a generic error type which we attempt to parse reverts into, to make things more readable
* A `revert_errors!` macro to define unions to handle transitive errors
* The state machine test instance is reused between runs, so we can cache Wasm modules in it
* `ArbTokenAmount` now generates values within the range of what Ethereum _and_ the FVM can handle, which is `U128` actually.
* Added a `IPC_ACTORS_NOFETCH` env var which when set to a value other than `0` allows us to make local changes to the Solidity actors, which then can be compiled with `make ipc-actors-abi`, and it won't check out the `IPC_ACTORS_TAG` from origin. (It only works if `scripts/find-ipc-actors.sh` already uses the fallback).

### CODE REVIEW

**Review by adlrocha — APPROVED** [review]
This is going to be super useful. As it is a long PR I may have missed some detail, let me know if you want me to have a deeper look to something before merging. Thanks!

File: fendermint/vm/interpreter/src/fvm/state/fevm.rs [hunk 1]
```diff
@@ -19,6 +22,55 @@ use super::FvmExecState;
 pub type MockProvider = ep::Provider<ep::MockProvider>;
 pub type MockContractCall<T> = ethers::prelude::ContractCall<MockProvider, T>;
 
+/// Result of trying to decode the data returned in failures as reverts.
+///
+/// The `E` type is supposed to be the enum unifying all errors that the contract can emit.
+pub enum CallError<E> {
```
[the remark this answers is not in the document]
Comment (aakoshh) [reply fendermint/vm/interpreter/src/fvm/state/fevm.rs:28]: You can't leverage this directly in your CLI, it demonstrates that you can parse the data from the output in the CLI the way I did here. I added another type called `SubnetError` in https://github.com/consensus-shipyard/fendermint/pull/299/commits/2ab532fd686824fb728c28052a324406ef6d0aba to help, which is something you _can_ use.

**Comment by aakoshh** [comment]
> Also CI seems quite unhappy :)

@adlrocha that's because the IPC agent is out of sync with the IPC Solidity actors and cannot be built:

```
error[E0560]: struct `ConstructorParams` has no field named `name`
   --> /home/runner/.cargo/git/checkouts/ipc-agent-31818396ce0bee3c/fd75ba4/ipc/provider/src/manager/evm/manager.rs:228:13
    |
228 |             name,
    |             ^^^^ `ConstructorParams` does not have this field
```

**Comment by adlrocha** [comment]
> @adlrocha that's because the IPC agent is out of sync with the IPC Solidity actors and cannot be built:

True, sorry, I didn't even look into what was the error. @cryptoAtwill's PRs have been merged, and I am hoping to get mine merged today (if all these meetings allow me to).

**Pull request merged**